File: frontend/src/types/developer.ts
/**
 * ARCH-21 — developer platform contracts.
 *
 * Mirrors app/schemas/developer.py and app/schemas/public_api.py.
 *
 * EVERY LATENCY AND RATE FIELD IS `number | null`, NOT `number`.
 * ==============================================================
 *
 * `null` means "no measurement", and the components in
 * OrganizationDeveloperPortal.tsx render it as an em dash. Typing these as
 * `number` would push the backend's honest null through `?? 0` at some call
 * site, and a window with no traffic would render as a service that responds
 * in zero milliseconds with a flawless error rate. That is the same
 * laundering of an unknown into a flattering number that ARCH-18 prohibited
 * when it banned COALESCE(cost_basis_micros, 0), and the type is what stops
 * it here.
 */

/** Mirrors app/core/api_tiers.py::ApiRateTier. */
export const API_RATE_TIERS = ["FREE", "BUILDER", "PRO", "ENTERPRISE"] as const;

export type ApiRateTier = (typeof API_RATE_TIERS)[number];

/** Mirrors app/core/scopes.py::PUBLIC_API_SCOPES. */
export const PUBLIC_API_SCOPES = [
  "public_documents:read",
  "public_query:write",
  "public_workflows:read",
  "public_workflows:write",
] as const;

export type PublicApiScope = (typeof PUBLIC_API_SCOPES)[number];

export interface ApiTierDescriptor {
  readonly key: ApiRateTier;
  readonly display_name: string;
  readonly rank: number;
  readonly rate_limit_per_minute: number;
  readonly monthly_request_quota: number;
  readonly ef_search: number;
  readonly description: string;
  /**
   * False when the tier is above the organization's plan ceiling. The tier is
   * still listed — hiding it would leave an admin unable to see that a higher
   * tier exists or what upgrading buys.
   */
  readonly assignable: boolean;
}

export interface TierCatalogue {
  readonly ceiling: ApiRateTier;
  /** null when no quota_tiers version is in force. Resolves the ceiling to FREE. */
  readonly quota_tier_key: string | null;
  readonly tiers: readonly ApiTierDescriptor[];
}

export interface DeveloperKeySummary {
  readonly id: string;
  readonly name: string;
  /** Computed from the key id. There is no prefix column; this carries no secret. */
  readonly display_prefix: string;
  readonly tier_key: ApiRateTier;
  readonly rate_limit_per_minute: number;
  readonly monthly_request_quota: number;
  readonly is_public_api_enabled: boolean;
  readonly scopes: readonly string[];
  readonly public_scopes: readonly string[];
  readonly expires_at: string | null;
  readonly last_used_at: string | null;
  readonly month_to_date_requests: number;
  readonly window_requests: number;
  /** null when the quota is zero — a fraction of nothing is undefined, not 0. */
  readonly quota_used_fraction: number | null;
  readonly created_at: string;
}

export interface DeveloperOverview {
  readonly organization_id: string;
  readonly window_days: number;
  readonly window_start: string;
  readonly window_end: string;
  readonly tier_catalogue: TierCatalogue;
  readonly keys: readonly DeveloperKeySummary[];
  readonly public_key_count: number;
  readonly total_key_count: number;
  readonly month_to_date_requests: number;
}

export interface DeveloperUsagePoint {
  readonly date: string;
  readonly request_count: number;
  readonly error_count: number;
  readonly throttled_count: number;
  readonly mean_latency_ms: number | null;
}

export interface DeveloperKeyMetrics {
  readonly api_key_id: string;
  readonly window_days: number;
  readonly window_start: string;
  readonly window_end: string;
  readonly total_requests: number;
  readonly total_errors: number;
  readonly total_throttled: number;
  readonly served_requests: number;
  readonly error_rate: number | null;
  readonly mean_latency_ms: number | null;
  readonly p50_latency_ms: number | null;
  readonly p95_latency_ms: number | null;
  /** Always HISTOGRAM_INTERPOLATED. Percentiles are estimates to one bucket width. */
  readonly latency_method: string;
  readonly month_to_date_requests: number;
  readonly series: readonly DeveloperUsagePoint[];
}

export interface DeveloperTierUpdateRequest {
  readonly tier_key: ApiRateTier;
  readonly enable_public_api?: boolean;
}

export interface DeveloperKeyCreateRequest {
  readonly name: string;
  readonly scopes: readonly string[];
  readonly tier_key: ApiRateTier;
  readonly expires_at?: string | null;
  readonly enable_public_api: boolean;
}

export interface DeveloperKeyIssued {
  readonly api_key: DeveloperKeySummary;
  /** Shown once. Never stored, never logged, never re-fetchable. */
  readonly token: string;
}

export interface CodeSnippetSet {
  readonly curl: string;
  readonly python: string;
  readonly typescript: string;
}

export type SnippetLanguage = keyof CodeSnippetSet;

export const SNIPPET_LANGUAGES: readonly {
  readonly id: SnippetLanguage;
  readonly label: string;
}[] = [
  { id: "curl", label: "cURL" },
  { id: "python", label: "Python" },
  { id: "typescript", label: "TypeScript" },
];

export interface ApiExplorerOperation {
  readonly operation_id: string;
  readonly method: string;
  readonly path: string;
  readonly summary: string;
  readonly required_scope: string;
  readonly snippets: CodeSnippetSet;
}

export interface ApiExplorerCatalogue {
  readonly base_url: string;
  readonly operations: readonly ApiExplorerOperation[];
}

/* ------------------------------------------------------------------------ */
/* Formatting helpers                                                        */
/* ------------------------------------------------------------------------ */

/**
 * The one place a null becomes visible text.
 *
 * Centralised so no component reaches for `?? 0`. If a number is absent the
 * user sees an em dash and the tooltip explains why; they never see a zero
 * that was never measured.
 */
export const formatMeasurement = (
  value: number | null | undefined,
  unit: string,
  fractionDigits = 0,
): string => {
  if (value === null || value === undefined || Number.isNaN(value)) {
    return "—";
  }
  return `${value.toFixed(fractionDigits)}${unit}`;
};

export const formatPercent = (
  fraction: number | null | undefined,
  fractionDigits = 1,
): string => {
  if (fraction === null || fraction === undefined || Number.isNaN(fraction)) {
    return "—";
  }
  return `${(fraction * 100).toFixed(fractionDigits)}%`;
};

export const formatCount = (value: number): string =>
  new Intl.NumberFormat(undefined, { notation: "compact" }).format(value);

/**
 * Quota bar width. Returns null rather than 0 when the fraction is unknown,
 * so the caller renders an indeterminate bar instead of an empty one — an
 * empty bar reads as "nothing used", which is a claim.
 */
export const quotaBarFraction = (
  key: Pick<DeveloperKeySummary, "quota_used_fraction">,
): number | null => {
  const value = key.quota_used_fraction;
  if (value === null || value === undefined || Number.isNaN(value)) {
    return null;
  }
  return Math.max(0, Math.min(1, value));
};

export const isTierAssignable = (
  catalogue: TierCatalogue | undefined,
  tier: ApiRateTier,
): boolean =>
  catalogue?.tiers.find((entry) => entry.key === tier)?.assignable ?? false;

File: backend/app/schemas/developer.py
"""ARCH-21 §4.4 — developer portal contracts.

Every latency field here is Optional and every rate field is Optional, and
that is the single most important thing about this module. `None` means "no
measurement", and the frontend renders it as such. A schema that declared
`p95_latency_ms: float = 0.0` would turn a day with no traffic into a service
that responds in zero milliseconds — the same laundering of an unknown into a
flattering number that ARCH-18 banned when it prohibited
`COALESCE(cost_basis_micros, 0)`.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.core.api_tiers import ApiRateTier
from app.core.scopes import ApiKeyScope


class ApiTierDescriptor(BaseModel):
    key: str
    display_name: str
    rank: int
    rate_limit_per_minute: int
    monthly_request_quota: int
    ef_search: int
    description: str
    assignable: bool = Field(
        description=(
            "False when this tier is above the organization's plan ceiling. "
            "Listed anyway so an admin can see what upgrading would buy."
        )
    )


class TierCataloguePayload(BaseModel):
    ceiling: str
    quota_tier_key: Optional[str] = Field(
        default=None,
        description=(
            "The organization's commercial quota tier. None means no "
            "quota_tiers version is in force, which resolves the ceiling to "
            "FREE rather than to unlimited."
        ),
    )
    tiers: list[ApiTierDescriptor]


class DeveloperKeySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    display_prefix: str = Field(
        description=(
            "Computed from the key id, never stored. There is no prefix "
            "column on api_keys; this is the id in base32 behind the "
            "environment tag, and it carries no secret."
        )
    )
    tier_key: str
    rate_limit_per_minute: int
    monthly_request_quota: int
    is_public_api_enabled: bool
    scopes: list[str]
    public_scopes: list[str]
    expires_at: Optional[str] = None
    last_used_at: Optional[str] = None
    month_to_date_requests: int
    window_requests: int
    quota_used_fraction: Optional[float] = Field(
        default=None,
        description="None when the quota is zero; a fraction is undefined, not 0.",
    )
    created_at: str


class DeveloperOverviewResponse(BaseModel):
    organization_id: str
    window_days: int
    window_start: str
    window_end: str
    tier_catalogue: TierCataloguePayload
    keys: list[DeveloperKeySummary]
    public_key_count: int
    total_key_count: int
    month_to_date_requests: int


class DeveloperUsagePoint(BaseModel):
    date: str
    request_count: int
    error_count: int
    throttled_count: int
    mean_latency_ms: Optional[float] = None


class DeveloperKeyMetricsResponse(BaseModel):
    api_key_id: str
    window_days: int
    window_start: str
    window_end: str
    total_requests: int
    total_errors: int
    total_throttled: int
    served_requests: int
    error_rate: Optional[float] = None
    mean_latency_ms: Optional[float] = None
    p50_latency_ms: Optional[float] = None
    p95_latency_ms: Optional[float] = None
    latency_method: str = Field(
        description=(
            "HISTOGRAM_INTERPOLATED. Percentiles are interpolated from a "
            "bucketed distribution and are accurate to one bucket width. "
            "They are estimates and are labelled as such."
        )
    )
    month_to_date_requests: int
    series: list[DeveloperUsagePoint]


class DeveloperTierUpdateRequest(BaseModel):
    tier_key: ApiRateTier
    enable_public_api: Optional[bool] = Field(
        default=None,
        description=(
            "Omit to leave the gateway opt-in unchanged. Setting true "
            "requires the key to already hold at least one public_* scope."
        ),
    )


class DeveloperKeyCreateRequest(BaseModel):
    """Issue a key and stamp it in one call.

    Separate from `ApiKeyCreate` because the console's issuance flow has no
    concept of a tier and must not acquire one: a key minted for an internal
    integration should stay gateway-disabled by default, which is exactly
    what happens when `is_public_api_enabled` defaults to false.
    """

    name: str = Field(min_length=1, max_length=120)
    scopes: list[ApiKeyScope] = Field(min_length=1)
    tier_key: ApiRateTier = ApiRateTier.FREE
    expires_at: Optional[datetime] = None
    enable_public_api: bool = False


class DeveloperKeyIssuedResponse(BaseModel):
    api_key: DeveloperKeySummary
    token: str = Field(
        description="Full token. Shown ONCE, never stored, never logged."
    )


class CodeSnippetSet(BaseModel):
    curl: str
    python: str
    typescript: str


class ApiExplorerOperation(BaseModel):
    operation_id: str
    method: str
    path: str
    summary: str
    required_scope: str
    snippets: CodeSnippetSet


class ApiExplorerResponse(BaseModel):
    base_url: str
    operations: list[ApiExplorerOperation]


__all__ = [
    "ApiExplorerOperation",
    "ApiExplorerResponse",
    "ApiTierDescriptor",
    "CodeSnippetSet",
    "DeveloperKeyCreateRequest",
    "DeveloperKeyIssuedResponse",
    "DeveloperKeyMetricsResponse",
    "DeveloperKeySummary",
    "DeveloperOverviewResponse",
    "DeveloperTierUpdateRequest",
    "DeveloperUsagePoint",
    "TierCataloguePayload",
]